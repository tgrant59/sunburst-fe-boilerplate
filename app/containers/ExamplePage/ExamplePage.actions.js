export const LOAD_USER_REQUESTED =
    'fe-app-boilerplate/ExamplePage/LOAD_USER_REQUESTED'
export const LOAD_USER_SUCCESS =
    'fe-app-boilerplate/ExamplePage/LOAD_USER_SUCCESS'
export const LOAD_USER_FAILED =
    'fe-app-boilerplate/ExamplePage/LOAD_USER_FAILED'

export const loadUser = () => ({
    type: LOAD_USER_REQUESTED,
})
export const loadUserSuccess = currentUserData => ({
    type: LOAD_USER_SUCCESS,
    payload: { currentUserData },
})
export const loadUserFail = error => ({
    type: LOAD_USER_FAILED,
    payload: { error },
})
