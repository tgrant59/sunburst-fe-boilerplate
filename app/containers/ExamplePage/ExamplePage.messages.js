export default {
    helmet: {
        title: 'Users',
        description: 'The user example page of Top Hat',
    },
    welcomeMessage: 'Hello, World!',
    errorLoadingUser: 'Error Loading User',
    loadingUser: 'Loading User...',
}
