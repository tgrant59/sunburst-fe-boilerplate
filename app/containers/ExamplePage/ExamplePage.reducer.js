import { fromJS } from 'immutable'

import {
    LOAD_USER_REQUESTED,
    LOAD_USER_SUCCESS,
    LOAD_USER_FAILED,
} from './ExamplePage.actions'

export const initialState = fromJS({
    currentUser: null,
    currentUserError: false,
})

const exampleReducer = (state = initialState, action) => {
    switch (action.type) {
        case LOAD_USER_REQUESTED:
            return state.merge({
                currentUser: null,
                currentUserError: false,
            })
        case LOAD_USER_SUCCESS:
            return state.merge({
                currentUser: action.payload.currentUserData,
                currentUserError: false,
            })
        case LOAD_USER_FAILED:
            return state.merge({
                currentUserError: true,
            })
        default:
            return state
    }
}

export default exampleReducer
