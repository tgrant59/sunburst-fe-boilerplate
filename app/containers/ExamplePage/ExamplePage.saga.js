import { call, put, takeLatest } from 'redux-saga/effects'

import request from 'helpers/request'

import {
    LOAD_USER_REQUESTED,
    loadUserFail,
    loadUserSuccess,
} from './ExamplePage.actions'

export function* getUserData() {
    const requestURL = `https://randomuser.me/api/`

    try {
        const response = yield call(request.get, requestURL)
        const userData = response.data.results[0]
        yield put(
            loadUserSuccess({
                name: `${userData.name.first} ${userData.name.last}`,
                email: userData.email,
            }),
        )
    } catch (error) {
        yield put(loadUserFail(error.response.status))
    }
}

export default function* exampleSagaWatcher() {
    yield takeLatest(LOAD_USER_REQUESTED, getUserData)
}
