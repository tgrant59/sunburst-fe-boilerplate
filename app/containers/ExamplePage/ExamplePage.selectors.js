import { createSelector } from 'reselect'

export const selectExampleState = state => state.get('example')

export const selectCurrentUser = createSelector(
    selectExampleState,
    exampleState => {
        if (!exampleState) return null
        return exampleState.get('currentUser')
    },
)

export const selectCurrentUserError = createSelector(
    selectExampleState,
    exampleState => {
        if (!exampleState) return null
        return exampleState.get('currentUserError')
    },
)
