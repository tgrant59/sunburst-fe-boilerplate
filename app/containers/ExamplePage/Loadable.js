import Loadable from 'react-loadable'

import PageLoading from 'components/PageLoading'

export default Loadable({
    loader: () => import('./index'),
    loading: PageLoading,
})
