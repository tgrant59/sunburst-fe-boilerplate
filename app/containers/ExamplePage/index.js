import React from 'react'
import PropTypes from 'prop-types'
import { Helmet } from 'react-helmet'
import { connect } from 'react-redux'
import { compose } from 'redux'
import styled from 'styled-components'

import injectReducer from 'utils/injectReducer'
import injectSaga from 'utils/injectSaga'

import { loadUser } from './ExamplePage.actions'
import messages from './ExamplePage.messages'
import exampleReducer from './ExamplePage.reducer'
import exampleSaga from './ExamplePage.saga'
import {
    selectCurrentUser,
    selectCurrentUserError,
} from './ExamplePage.selectors'

const CenteredDiv = styled.div`
    text-align: center;
`

// eslint-disable-next-line react/prefer-stateless-function
export class ExamplePage extends React.PureComponent {
    static propTypes = {
        currentUser: PropTypes.shape({
            name: PropTypes.string,
            email: PropTypes.string,
        }),
        currentUserError: PropTypes.bool,
        loadUser: PropTypes.func.isRequired,
    }

    static defaultProps = {
        currentUserError: false,
    }

    componentDidMount() {
        this.props.loadUser()
    }

    getUserData() {
        const user = this.props.currentUser
        if (this.props.currentUserError) {
            return <p>{messages.errorLoadingUser}</p>
        } else if (user) {
            return (
                <p>
                    {user.get('name')} - {user.get('email')}
                </p>
            )
        }
        return <p>{messages.loadingUser}</p>
    }

    render() {
        return (
            <CenteredDiv>
                <Helmet>
                    <title>{messages.helmet.title}</title>
                    <meta
                        name="description"
                        content={messages.helmet.description}
                    />
                </Helmet>
                <h1>{messages.welcomeMessage}</h1>
                {this.getUserData()}
            </CenteredDiv>
        )
    }
}

export const mapStateToProps = state => ({
    currentUser: selectCurrentUser(state),
    currentUserError: selectCurrentUserError(state),
})

export const mapDispatchToProps = {
    loadUser,
}

const withState = compose(
    injectReducer({ key: 'example', reducer: exampleReducer }),
    injectSaga({ key: 'example', saga: exampleSaga }),
    connect(mapStateToProps, mapDispatchToProps),
)

export default withState(ExamplePage)
