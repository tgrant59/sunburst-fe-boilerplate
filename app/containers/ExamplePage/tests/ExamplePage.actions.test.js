import { loadUser, loadUserFail, loadUserSuccess } from '../ExamplePage.actions'
import { mockCurrentUser } from './mocks'

describe('User Actions', () => {
    describe('loadUser', () => {
        it('should match snapshot', () => {
            expect(loadUser()).toMatchSnapshot()
        })
    })

    describe('loadUserSuccess', () => {
        it('should match snapshot', () => {
            expect(loadUserSuccess(mockCurrentUser)).toMatchSnapshot()
        })
    })

    describe('loadUserFail', () => {
        it('should match snapshot', () => {
            expect(loadUserFail()).toMatchSnapshot()
        })
    })
})
