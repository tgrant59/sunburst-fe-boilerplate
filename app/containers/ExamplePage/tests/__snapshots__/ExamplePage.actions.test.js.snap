// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`User Actions loadUser should match snapshot 1`] = `
Object {
  "type": "fe-app-boilerplate/ExamplePage/LOAD_USER_REQUESTED",
}
`;

exports[`User Actions loadUserFail should match snapshot 1`] = `
Object {
  "payload": Object {
    "error": undefined,
  },
  "type": "fe-app-boilerplate/ExamplePage/LOAD_USER_FAILED",
}
`;

exports[`User Actions loadUserSuccess should match snapshot 1`] = `
Object {
  "payload": Object {
    "currentUserData": Object {
      "email": "jer@yolo.net",
      "name": "Jeremy Sant",
    },
  },
  "type": "fe-app-boilerplate/ExamplePage/LOAD_USER_SUCCESS",
}
`;
